pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/impl/LoadManagerShared.java

```diff
@@ -409,22 +413,23 @@ private static void filterDomainsNotHavingLeastNumberAntiAffinityNamespaces(
             return nsCount != null && nsCount != finalLeastNamespaceCount;
         });
     }
-
+    @FunctionalInterface
+    private interface CountAntiAffinityNamespaceOwnedBrokers{
+        void run(String broker, String namespace);
+    }
     /**
      * It returns map of broker and count of namespace that are belong to the same anti-affinity group as given.
      *
      * @param pulsar
      * @param namespaceName
-     * @param brokerToNamespaceToBundleRange
+     * @param bundleOwnershipData
      * @return
      */
     public static CompletableFuture<Map<String, Integer>> getAntiAffinityNamespaceOwnedBrokers(
```

Comment:
Can we change the `bundleOwnershipData` to `Supplier<Map<String, String>> brokerToNsMapSupplier`?
```java
    public static CompletableFuture<Map<String, Integer>> getAntiAffinityNamespaceOwnedBrokers(
            final PulsarService pulsar, final String namespaceName,
            Supplier<Map<String, String>> brokerToNsMapSupplier) {

        CompletableFuture<Map<String, Integer>> antiAffinityNsBrokersResult = new CompletableFuture<>();
        pulsar.getPulsarResources().getLocalPolicies().getLocalPoliciesAsync(NamespaceName.get(namespaceName))
                .thenAccept(policies -> {
            if (!policies.isPresent() || StringUtils.isBlank(policies.get().namespaceAntiAffinityGroup)) {
                antiAffinityNsBrokersResult.complete(null);
                return;
            }
            final String antiAffinityGroup = policies.get().namespaceAntiAffinityGroup;
            final Map<String, Integer> brokerToAntiAffinityNamespaceCount = new ConcurrentHashMap<>();
            final List<CompletableFuture<Void>> futures = new ArrayList<>();

            Map<String, String> brokerToNsMap = brokerToNsMapSupplier.get();
            brokerToNsMap.forEach((broker, ns) -> {
                CompletableFuture<Void> future = new CompletableFuture<>();
                futures.add(future);

                pulsar.getPulsarResources().getLocalPolicies().getLocalPoliciesAsync(NamespaceName.get(ns))
                        .thenAccept(nsPolicies -> {
                            if (nsPolicies.isPresent()
                                    && antiAffinityGroup.equalsIgnoreCase(nsPolicies.get().namespaceAntiAffinityGroup)) {
                                brokerToAntiAffinityNamespaceCount.compute(broker,
                                        (brokerName, count) -> count == null ? 1 : count + 1);
                            }
                            future.complete(null);
                        }).exceptionally(ex -> {
                            future.complete(null);
                            return null;
                        });
            });
            FutureUtil.waitForAll(futures)
                    .thenAccept(r -> antiAffinityNsBrokersResult.complete(brokerToAntiAffinityNamespaceCount));
        }).exceptionally(ex -> {
            // namespace-policies has not been created yet
            antiAffinityNsBrokersResult.complete(null);
            return null;
        });
        return antiAffinityNsBrokersResult;
    }
```

Reply:
brokerToNamespaceToBundleRange is a nested HashMap. 
If we do that, we need to flatten this nested map first, which seems unnecessary.

Reply:
Also, ConcurrentOpenHashMap does not implement java.util.Map(it is a custom HashMap implementation)

Reply:
It's confusing to use an Object type param. Actually, the method needs a list of data pairs (broker -> namespace), maybe the caller can filter invalid data, and prepare the data first.

For `ModularLoadManagerImpl`, convert `ConcurrentOpenHashMap` to a List, and filter empty bundle ranges, then call this method.
For `ExtensibleLoadManagerImpl.java`, convert `Set` to a List, and filter bundles with an invalid state, then call this method.

Reply:
It's confusing to add an Object type param. Actually, the method needs a list of data pairs (broker -> namespace), maybe the caller can filter invalid data, and prepare the data first.

For `ModularLoadManagerImpl`, convert `ConcurrentOpenHashMap` to a List, and filter empty bundle ranges, then call this method.
For `ExtensibleLoadManagerImpl`, convert `Set` to a List, filter bundles with an invalid state, then call this method.